### PR TITLE
rebased to upstream tag REL8_4_22

### PR DESCRIPTION
**Pull request by davecramer**

### CODE REVIEW

**Comment by Eulerizeit** [comment]
@schubert @pivotal-mike FYI first one down.

**Comment by pivotal-mike** [comment]
+1

**Comment by hlinnaka** [comment]
Ok, why? I must be missing some context here.

File: contrib/pgcrypto/pgp-decrypt.c [hunk 1]
```diff
@@ -1187,36 +1182,8 @@ pgp_decrypt(PGP_Context *ctx, MBuf *msrc, MBuf *mdst)
...
-	/*
-	 * Report a failure of the prefix_init() "quick check" now, rather than
-	 * upon detection, to hinder timing attacks.  pgcrypto is not generally
-	 * secure against timing attacks, but this helps.
-	 */
 	if (!got_data || ctx->corrupt_prefix)
-		return PXE_PGP_CORRUPT_DATA;
-
-	/*
-	 * Code interpreting purportedly-decrypted data prior to this stage shall
```
Comment (hlinnaka) [inline contrib/pgcrypto/pgp-decrypt.c]: This code that you're removing was a backport of this upstream commit:
```
commit 7b758b7d605aca10b36aa1c26bbf16c33f8ac726
Author: Noah Misch <noah@leadboat.com>
Date:   Mon May 18 10:02:31 2015 -0400

    pgcrypto: Report errant decryption as "Wrong key or corrupt data".
    
    This has been the predominant outcome.  When the output of decrypting
    with a wrong key coincidentally resembled an OpenPGP packet header,
    pgcrypto could instead report "Corrupt data", "Not text data" or
    "Unsupported compression algorithm".  The distinct "Corrupt data"
    message added no value.  The latter two error messages misled when the
    decrypted payload also exhibited fundamental integrity problems.  Worse,
    error message variance in other systems has enabled cryptologic attacks;
    see RFC 4880 section "14. Security Considerations".  Whether these
    pgcrypto behaviors are likewise exploitable is unknown.
    
    In passing, document that pgcrypto does not resist side-channel attacks.
    Back-patch to 9.0 (all supported versions).
    
    Security: CVE-2015-3167
```
We don't want to lose that.

Comment (schubert) [inline contrib/pgcrypto/pgp-decrypt.c]: Right so the missing bit of info here for folks is being careful of pulling upstream nuking future backports that were done intentionally - in this case a CVE which are very critical to not lose. We can't regress on security issues.

Comment (schubert) [inline contrib/pgcrypto/pgp-decrypt.c]: @hlinnaka this raises an important question: how does anyone identify what might have been a future back port? Do we have any way to tell what commits beyond 8.3 were pulled in for reasons like this (the CVE fix)?

File: contrib/pgcrypto/init_file [hunk 2]
```diff
@@ -0,0 +1,88 @@
+-- This file contains global patterns of messages that should be ignored or
```
Comment (hlinnaka) [inline contrib/pgcrypto/init_file]: Can you trim this down to only the stuff that's actually needed by pgcrypto, please? I assume that's only the "'DISTRIBUTED BY' clause" rule

File: contrib/pgcrypto/expected/pgp-pubkey-decrypt.out [hunk 3]
```diff
@@ -9,10 +9,14 @@ create table keytbl (
 	pubkey text,
 	seckey text
 );
+NOTICE:  Table doesn't have 'DISTRIBUTED BY' clause -- Using column named 'id' as the Greenplum Database data distribution key for this table.
```
Comment (hlinnaka) [inline contrib/pgcrypto/expected/pgp-pubkey-decrypt.out]: If you're adding suppression for these in the init_file, no need to also include them in the expected output.

File: contrib/pgcrypto/expected/pgp-pubkey-decrypt.out [hunk 4]
```diff
@@ -617,15 +621,15 @@ from keytbl, encdata where keytbl.id=6 and encdata.id=4;
 -- wrong key
 select pgp_pub_decrypt(dearmor(data), dearmor(seckey))
 from keytbl, encdata where keytbl.id=2 and encdata.id=1;
-ERROR:  Wrong key
+ERROR:  Wrong key  (seg0 slice2 127.0.1.1:25435 pid=17723)
```
Comment (hlinnaka) [inline contrib/pgcrypto/expected/pgp-pubkey-decrypt.out]: let's not change the expected output unnecessarily

File: contrib/pgcrypto/expected/pgp-pubkey-decrypt.out [hunk 5]
```diff
@@ -637,20 +641,16 @@ from keytbl, encdata where keytbl.id=7 and encdata.id=4;
 -- password-protected secret key, no password
 select pgp_pub_decrypt(dearmor(data), dearmor(seckey))
 from keytbl, encdata where keytbl.id=5 and encdata.id=1;
-ERROR:  Need password for secret key
+ERROR:  Need password for secret key  (seg0 slice2 127.0.1.1:25435 pid=17723)
 -- password-protected secret key, wrong password
 select pgp_pub_decrypt(dearmor(data), dearmor(seckey), 'foo')
 from keytbl, encdata where keytbl.id=5 and encdata.id=1;
-ERROR:  Wrong key or corrupt data
+ERROR:  Unsupported cipher algorithm  (seg0 slice2 127.0.1.1:25435 pid=17723)
```
Comment (hlinnaka) [inline contrib/pgcrypto/expected/pgp-pubkey-decrypt.out]: Where did this change in output come from? I'm not seeing this on my laptop. Same with the next change.

**Pull request merged**